internal/messagequeue/messagequeue.go

```diff
@@ -327,6 +357,61 @@ func (mq *MessageQueue) sendMessage() {
 	}
 }
 
+// If the peer is running an older version of Bitswap that doesn't support the
+// DONT_HAVE response, watch for timeouts on any want-blocks we sent the peer,
+// and if there is a timeout simulate a DONT_HAVE response.
+func (mq *MessageQueue) simulateDontHaveWithTimeout(msg bsmsg.BitSwapMessage) {
+	// If the peer supports DONT_HAVE responses, we don't need to simulate
+	if mq.sender.SupportsHave() {
+		return
+	}
+
+	mq.wllock.Lock()
+	defer mq.wllock.Unlock()
+
+	// Get the CID of each want-block that expects a DONT_HAVE response
+	wantlist := msg.Wantlist()
+	wants := make([]cid.Cid, 0, len(wantlist))
+	for _, entry := range wantlist {
+		if entry.WantType == pb.Message_Wantlist_Block && entry.SendDontHave {
+			// Unlikely, but just in case check that the block hasn't been
+			// received in the interim
+			c := entry.Cid
+			if _, ok := mq.peerWants.allWants.Contains(c); ok {
+				// Check we don't already have a timer running for this cid
+				if _, existing := mq.dontHaveTimers[c]; !existing {
+					wants = append(wants, c)
+				}
+			}
+		}
+	}
+
+	if len(wants) == 0 {
+		return
+	}
+
+	// Create a timer that will fire if we don't receive a block
```

Comment:
@Stebalien what do you think of this approach of using timers?
Is it likely to be performant enough or should we do something coarser grained, like run our own tick function every X milliseconds that looks for expired wants?

Reply:
Profiling it would be best. Given that we're doing it once per _message_ and not per _cid_, we might actually be fine (we already tend to allocate new timers whenever we set a deadline on a context).

Otherwise, we could probably just stick CIDs into a queue with a ticker as you say.

Reply:
Really, let's leave this for now and see what happens.